src/storage.ts

```diff
@@ -0,0 +1,266 @@
+// Storage utility types
+enum StorageAccessLevel {
+  Public = 'public',
+  Private = 'private',
+  Protected = 'protected'
+};
+
+enum PageSize {
+  All = 'ALL'
+};
+
+enum EncryptionOption {
+  AES256 = 'AES256',
+  KMS = 'aws:kms'
+}
+
+enum ACLOption {
+  AuthenticatedRead = 'authenticated-read',
+  AWSExecRead = 'aws-exec-read',
+  BucketOwnerFullControl = 'bucket-owner-full-control',
+  BucketOwnerRead = 'bucket-owner-read',
+  Private = 'private',
+  PublicRead = 'public-read',
+  PublicReadWrite = 'public-read-write'
+}
+
+type TransferProgress = {
+  readonly transferred: number;
+  readonly total: number;
+};
+
+type AccessLevelConfig = {
+  readonly accessLevel?: StorageAccessLevel;
+  readonly identityId?: string;
+}
+
+type StorageObjectReference = {
+  readonly key: string;
+  readonly size?: number;
+  readonly eTag?: string;
```

Comment:
how the customer will generate this? I saw this on `put` function

Reply:
Jim and I have talked and I'm recommending something like the following as the input type for public interfaces
```
type PartialStorageObjectReference = string | Partial<Omit<FileReference, 'key'> | Require<FileReference, 'key'>;
```

From this, internally we can fill in defaults to create a fully vetted `StorageObjectReference`, which we can then use to access/operate on/create an object.

I'm also recommending `StoragePrefixReference` (or Path) with a related Partial.

Reply:
Also recommend splitting out `Reference` (where is the object) and `Metadata` (what is the object), so that when we mean both we can union the types together, but we don't have extra empty Metadata fields in all the cases where we don't need them.